Restrict the main element to be used once per document

annevk opened the pull request:
Multiple main elements are still allowed if all but one have the hidden attribute set.

This also slightly alters the definition of the body element as it effectively encompasses all content of a document, except for some metadata.

Fixes #100.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/3354/acknowledgements.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">/acknowledgements.html</a>  ( <a href="https://whatpr.org/html/3354/97d8432...8200648/acknowledgements.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">diff</a> )
<a href="https://whatpr.org/html/3354/dom.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">/dom.html</a>  ( <a href="https://whatpr.org/html/3354/97d8432...8200648/dom.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">diff</a> )
<a href="https://whatpr.org/html/3354/grouping-content.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">/grouping-content.html</a>  ( <a href="https://whatpr.org/html/3354/97d8432...8200648/grouping-content.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">diff</a> )
<a href="https://whatpr.org/html/3354/indices.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">/indices.html</a>  ( <a href="https://whatpr.org/html/3354/97d8432...8200648/indices.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">diff</a> )
<a href="https://whatpr.org/html/3354/sections.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">/sections.html</a>  ( <a href="https://whatpr.org/html/3354/97d8432...8200648/sections.html" title="Last updated on Jan 19, 2018, 11:52 AM GMT (8200648)">diff</a> )

foolip commented:
@tobie, I think there's a differ bug here, outside of the changes made. Note how https://html.spec.whatwg.org/multipage/dom.html#phrasing-content and https://whatpr.org/html/3354/dom.html#phrasing-content both say "Text, in the context of content models" but in https://whatpr.org/html/3354/97d8432...38465fa/dom.html it's "Text , in the context of content models" with more whitespace. Seems like this is the case in lots of other places too. Can you take a look?

tobie commented:
@foolip: yeak that's an unfortunate but known issue of the [HTML diff service](https://services.w3.org/htmldiff) that pr-preview uses. It adds line breaks all over the place, which obviously creates issues with inline elements.

foolip commented on source:
```diff
@@ -16118,7 +16121,8 @@ Space is not the only void</pre>
    <dd>Where <span>flow content</span> is expected.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Flow content</span>, but with no <code>header</code>, <code>footer</code>, or <code>main</code> element descendants.</dd>
+   <dd><span>Flow content excluding main</span>, but with no <code>header</code> or
```
I'd prefer "Flow content excluding main" to be expanded in place, treating main more like other elements. Here: "Flow content, but with no header, footer or main element descendants." (unchanged, I think)

annevk replied:
I started with that and expected the review feedback to go the other way given that repeating yourself too much is never great. Happy to flip it back though.

foolip replied:
I may well have asked for the opposite if you did, who knows :)

I think the changes will be fewer if we trim the cases where main is disallowed, but let's start with figuring that out, and then inline the changes if the changes are no more than currently.

foolip commented on source:
```diff
@@ -42744,7 +42779,8 @@ interface <dfn>HTMLTableCellElement</dfn> : <span>HTMLElement</span> {
    <dd>Where <span>flow content</span> is expected.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Flow content</span>, but with no <code>form</code> element descendants.</dd>
+   <dd><span>Flow content excluding main</span>, but with no <code>form</code> element
```
This is the `form` element, any reason to change this?

annevk replied:
For the same reason I changed `td` and `blockquote`. If there's one main in a document it doesn't make sense as a descendant of them.

annevk commented on source:
```diff
@@ -15172,7 +15175,7 @@ interface <dfn>HTMLBodyElement</dfn> : <span>HTMLElement</span> {};
   </dl>
 
-  <p>The <code>body</code> element <span>represents</span> the main content of the document.</p>
+  <p>The <code>body</code> element <span>represents</span> the contents of the document.</p>
```
Yeah, I haven't actually looked at when this was written, but that seems correct.

annevk replied:
It seems this description of body goes back to 2006, before the `main` element was introduced.

foolip replied:
Thanks for checking that.

annevk commented on source:
```diff
@@ -19190,22 +19198,19 @@ included with Exhibit B.
 
-  <p class="note">While there is no restriction as to the number of <code>main</code> elements in a
-  document, web developers are encouraged to stick to a single element.</p>
+  <p>A document must not have more than one <code>main</code> element, unless all but one of the
```
I think must is preferable if you can, but sometimes the requirement might be hard to fulfill and then should can be reasonable.

foolip replied:
"must" wins in number of occurrences, so OK.

annevk commented:
Given where `main` makes sense out of these elements it seems it basically has to be a child of either `body` or `div`. Given that it doesn't make sense as being "flow content". I've updated this PR as appropriate. While doing so I also found that the indexes needed updating too.

stevefaulkner commented:
@annevk, `section` not an oversight, as there is no rule to stop use of a `section` which effectively contains everything in the document it seemed reasonable to allow in this case. Also given `section`s practical meaning as = `div`, it seemed OK to allow.   As for `blockquote` and 'td' yes they should be included in restrictions.

foolip commented on source:
```diff
@@ -15142,7 +15140,8 @@ interface <dfn>HTMLStyleElement</dfn> : <span>HTMLElement</span> {
    <dd>As the second element in an <code>html</code> element.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Flow content</span>.</dd>
+   <dd>Zero or one <code>main</code> element optionally intermixed with <span>flow
```
Should be zero or more to allow for `<main hidden>`, right?

foolip commented:
@tobie, seems like something's wrong with https://whatpr.org/html/3354/97d8432...8f1b75f/dom.html, commit 8f1b75f reverted "Flow content excluding main" but it's shown as an addition.

foolip commented:
Since we're down to debating what should happen with each case where flow content is allowed, here an exhaustive listing of the elements where "content model" has "flow content" in its description, and the validity of `<main>` within:
<table>
<tr><th>element</th><th>was allowed</th><th>now allowed</th>
<tr><td>address</td><td>yes</td><td>no (change)</td></tr>
<tr><td>article</td><td>yes</td><td>no (change)</td></tr>
<tr><td>aside</td><td>no</td><td>no</td></tr>
<tr><td>blockquote</td><td>yes</td><td>no (change)</td></tr>
<tr><td>body</td><td>yes</td><td>yes</td></tr>
<tr><td>caption</td><td>yes</td><td>no (change)</td></tr>
<tr><td>dd</td><td>yes</td><td>no (change)</td></tr>
<tr><td>details</td><td>yes</td><td>no (change)</td></tr>
<tr><td>dialog</td><td>yes</td><td>no (change)</td></tr>
<tr><td>div</td><td>yes</td><td>yes</td></tr>
<tr><td>dt</td><td>yes</td><td>no (change)</td></tr>
<tr><td>fieldset</td><td>yes</td><td>no (change)</td></tr>
<tr><td>figcaption</td><td>yes</td><td>no (change)</td></tr>
<tr><td>figure</td><td>yes</td><td>no (change)</td></tr>
<tr><td>footer</td><td>no</td><td>no</td></tr>
<tr><td>form</td><td>yes</td><td>no (change)</td></tr>
<tr><td>header</td><td>no</td><td>no</td></tr>
<tr><td>li</td><td>yes</td><td>no (change)</td></tr>
<tr><td>main</td><td>yes</td><td>no (changed)</td></tr>
<tr><td>nav</td><td>no</td><td>no</td></tr>
<tr><td>section</td><td>yes</td><td>no (change)</td></tr>
<tr><td>td</td><td>yes</td><td>no (change)</td></tr>
<tr><td>th</td><td>yes</td><td>no (change)</td></tr>
</table>

This is somewhat simplified because the content model of some elements, like `figure`, is more complicated.

tobie commented:
@foolip: OK, seems there's an AWS issue that's not being properly handled by pr-preview. Have you observed this elsewhere?

foolip commented:
@tobie, this is the first time I know that I've looked at a preview again after PR changes, so couldn't say if this has happened elsewhere.

annevk commented:
@stevefaulkner I think I'd prefer trying it out with restricting it from within `section` too as I think it makes little sense, but that's subject to review from @sideshowbarker.

Speaking of which, it'd be good to have an HTML checker PR lined up as well so when this change is made folks can verify the impact on their own sites.

stevefaulkner commented:
@annevk I am sanguine in regards to its restriction, as I recall it was restricted, but after discussion with Marcos Caceres I changed it to allow, but can't find the thread :-(

foolip commented:
I'd also like to hear from @sideshowbarker if there are some principles about validator warnings/errors we should keep in mind, or indeed principles about conformance generally.

As it is, I'm inclined to err on the side of allowing even things that seemingly don't make much sense unless we think they are both plausible mistakes and would cause user confusion/harm. Going for just body and div as the allowed parents, I'd like to take a look at what proportion of documents out there would become invalid. I don't want to go overboard with measuring and statistics, but it would be very bad if we made lots of harmless things invalid because we hadn't thought them plausible.

annevk commented:
I tried to think that through and I think requiring at least one without hidden is better, for progressive enhancement. Otherwise you'd end up with a mostly blank document if JavaScript is disabled, assuming `<main>` is used correctly to mark up the main contents.

foolip commented:
I don't think we should make any attempt to ensure that documents have some useful content with JavaScript disabled. Currently, `<body></body>` is valid, and so is `<table></table>`, both cases that don't make sense on their own.

annevk commented:
That is incorrect, we have a should requirement on that being not the case. See palpable content.

foolip commented:
https://html.spec.whatwg.org/multipage/dom.html#palpable-content says:
> As a general rule, elements whose content model allows any flow content or phrasing content should have at least one node in its contents that is palpable content and that does not have the hidden attribute specified.
>
> Note: Palpable content makes an element non-empty by providing either some descendant non-empty text, or else something users can hear (audio elements) or view (video or img or canvas elements) or otherwise interact with (for example, interactive form controls).
>
> This requirement is not a hard requirement, however, as there are many cases where an element can be empty legitimately, for example when it is used as a placeholder which will later be filled in by a script, or when the element is part of a template and would on most pages be filled in but on some pages is not relevant.
>
> Conformance checkers are encouraged to provide a mechanism for authors to find elements that fail to fulfill this requirement, as an authoring aid.

That's a lot of leeway for validators to point out things that might be wrong, and I'd really prefer to err on the side of allowing that which might be reasonable.

annevk commented:
> When a document has multiple main elements, at least all but one must have the hidden attribute specified. The remaining main element should not have the hidden attribute specified.
> 
> Note: this requirement is analogous to that for [palpable content].

sideshowbarker commented:
FYI, I’m in the process of adding use counters to the HTML checker backend to get some relevant data back on `main` usage — for example, the percentage of documents checked that contain `main` at all, any `main` with the `hidden` attribute, multiple unhidden `main`, any `main` within `section`, etc. 

I had hoped to get those use counters added today my time but it’s looking like I won’t until some time tomorrow (= 8+ hours from now).

foolip commented:
I've done some digging in httparchive to figure out what the ancestor chain for main elements normally looks like. Steps:
* [Looked in BigQuery](https://bigquery.cloud.google.com/savedquery/762219082167:ee0909fa67524c6180e13f1f2ddd1405) for response bodies that might have main elements.
* Saved those in [main.00.json](https://storage.googleapis.com/blink-httparchive-export/main/main.00.json) through [main.12.json](https://storage.googleapis.com/blink-httparchive-export/main/main.12.json)
* Wrote a [main.py](https://gist.github.com/foolip/53182eaf8c4dc434744b3b7364a8ac55) on all of them to print the ancestor chain
* Output is in [main.txt](https://gist.github.com/foolip/89d72f0575319f25c1082e41f9eed2d1) (, 1 per resource)

There are 51123 lines (resources). A bunch that say PARSEERROR and MAINCOUNT 0 cases were JS resources, because I didn't pre-filter for that. There were 48284 resource that had exactly 1 main element after parsing.

Here's the table from https://github.com/whatwg/html/pull/3354#issuecomment-357731086 with number of resources where the single main element have them on the ancestor chain:
<table>
<tr><th>element</th><th>count</th></tr>
<tr><td>address</td><td>0</td></tr>
<tr><td>article</td><td>56</td></tr>
<tr><td>aside</td><td>6</td></tr>
<tr><td>blockquote</td><td>1</td></tr>
<tr><td>body</td><td>48284</td></tr>
<tr><td>caption</td><td>0</td></tr>
<tr><td>dd</td><td>0</td></tr>
<tr><td>details</td><td>1</td></tr>
<tr><td>dialog</td><td>0</td></tr>
<tr><td>div</td><td>34977</td></tr>
<tr><td>dt</td><td>0</td></tr>
<tr><td>fieldset</td><td>1</td></tr>
<tr><td>figcaption</td><td>0</td></tr>
<tr><td>figure</td><td>3</td></tr>
<tr><td>footer</td><td>10</td></tr>
<tr><td>form</td><td>1162</td></tr>
<tr><td>header</td><td>51</td></tr>
<tr><td>li</td><td>20</td></tr>
<tr><td>main</td><td>3</td></tr>
<tr><td>nav</td><td>15</td></tr>
<tr><td>section</td><td>1601</td></tr>
<tr><td>td</td><td>36</td></tr>
<tr><td>th</td><td>0</td></tr>
</table>

Here are three cases each of form and section ancestors taken at random:
* http://www.redcross.ca/ (html body form div)
* http://www.gs1us.org/ (html body form div)
* http://www.edenred.gr/ (html body form div)
* http://www.uwosh.edu/ (html body div div section)
* http://www.roka.com/ (html body div section aside div)
* http://www.military-stuff.org/ (html body div div section)

The three form cases are all of the same kind, apparently using the same framework. @stevefaulkner, the role mapping of form seems to depend on the presence of "accessible name", can you tell if these cases would be confusing, or if wrapping the whole page in a form is harmless?

None of the section cases make a great deal of sense. uwosh.edu isn't using sections to mark up sections and roka.com isn't using main to mark up the dominant content, and main in aside is and will be invalid. military-stuff.org also uses section in an odd way, but is the least odd I think.

So, at least for section, I'm thinking that disallowing it might be a net positive.

foolip commented:
@sideshowbarker, have your counters revealed anything interesting?

annevk commented:
I ran those URLs through the https://checker.html5.org/ and the best case had 10 errors (but only because the last one was fatal and not recoverable). I don't think an additional error for this would be that bad.

sideshowbarker commented:
> @sideshowbarker, have your counters revealed anything interesting?

Along with the counters, I also set up some logging of URLs for documents that have instances of the cases that are getting counted. From that the (sorted and de-duped) logs are here:

https://gist.githubusercontent.com/sideshowbarker/8284404/raw/main-element.log

sideshowbarker commented:
@foolip, given https://github.com/whatwg/html/pull/3354#issuecomment-358604504, it seems maybe you already agree the cases other than `main` in `section` or `form` aren’t worth spending more time together on — so that is, for those other cases, the spec should define those as non-conforming (as @annevk wrote it already in this PR). Right?

Now as far as the `main` in `section` and `form` cases:

If I’m reading the httparchive data from https://github.com/whatwg/html/pull/3354#issuecomment-358604504 correctly, the 2.4% usage figure for `main` in `form` (1162 of 48284 total documents with `main`, right?) is significantly higher — 6 times higher— than what the HTML checker stats recorded (just 190 out of 47684 with `main`, so only 0.4%).

I’m curious about the reason for that discrepancy — but one thing I can say is, looking at it intuitively, I’d expect  to find `main` in `section` being used significantly more often than `main` in `form`. And when I look at the HTML checker stats, that is in fact what I find:  `main` in `section` is used about 5 times more often than `main` in `form` (and `main` in `section` itself isn’t even used that often to begin with: in only *0.3%* of the documents checked…).

So I dunno maybe that httparchive data needs to doublechecked to see why it unexpectedly shows `main` in `section` being used only 1.3 times as often as `main` in `form`.

Regardless, for the HTML checker we can say the numbers there show that documents processed by the HTML checker use `main` in `form` only very very rarely; specifically, in only 0.06% of documents.

And if that data’s anywhere close to accurate and `main` in `form` only appears that rarely in practice, then when an author comes across an instance of `main` in `form`, that means it’s probably more likely a mistake (unintended misuse, misplaced) than it is something the author actually intends to have there.

So in terms of the HTML checker behavior for something like that: I think emitting an error in that case may have a not-insignificant chance of helping an author catch a mistake they might have otherwise missed — a chance at least higher than the number of times that error is likely to end annoying anybody who’s *intentionally* using `main` in `form` for some considered reason.

> I'd also like to hear from @sideshowbarker if there are some principles about validator warnings/errors we should keep in mind, or indeed principles about conformance generally.

One general principle is what I tried to articulate above: If we know of some particular markup case that doesn’t seem clearly useful/logical — that is, we all can’t ourselves think of good cases where it would make sense, and we have no other evidence of it being put to good use by many people — then the spec should make that case non-conforming and the HTML checker should report an error for it.

The rationale behind that principle is what I also tried to articulate above: If a document contains such a markup case, it is more likely an unintended mistake than a willful usage.

And the purpose both for having document-conformance restrictions in the spec to begin with and for having the HTML checker report errors for them is to help authors catch those kinds of mistakes.

For a restatement of that, see https://validator.w3.org/nu/about.html#why-validate:

> The core reason to run your HTML documents through a conformance checker is simple: To catch unintended mistakes—mistakes you might have otherwise missed—so that you can fix them.

annevk commented:
Even though the data is a little different, the markup patterns seem very similar. `<form>` as a top-level element, wrapping `<header>`, `<main>`, etc. seems to be some kind of ASP.NET construct. It'd be great to hear from @whatwg/a11y / @stevefaulkner what the impact of that might be and whether allowing it is harmless. If it's harmless I'd be inclined to allow that.

stevefaulkner commented:
@annevk yes I think its needed for some ASP.net pages for some reason. Suggest that if form is used as a top-level element in this manner then it's OK to allow.

AliceWonderMiscreations commented:
On the hidden attribute thing, I have some web pages where there is a signle main attribute that is hidden until the user verifies they want to see the content (it is not safe for work).

The header and footer are not hidden, so someone visiting those pages can still find necessary contact information without needing to enable what they may not want to see.

So please don't require that there always be a main that doesn't have the hidden attribute. Otherwise I will have to break the spec to keep doing what I do, which is the ethically right thing to do (keeping potentially offensive content from display until user acknowledges)

iandevlin commented:
@stevefaulkner Yes, ASP.NET requires everything to be wrapped in a `<form>` because the entire framework revolves around POST commands.

stevefaulkner commented:
@AliceWonderMiscreations 

> So please don't require that there always be a main that doesn't have the hidden attribute.

I don't think it is a good idea to require a visible `main` element on every page, but was not aware that its was being considered as a requirement.

annevk commented:
See https://github.com/whatwg/html/pull/3354#issuecomment-357879900 for an alternative suggestion that makes it a "should not", but given the use case it seems reasonable to just allow it as @foolip has been suggesting all along.

foolip commented:
@sideshowbarker 

> @foolip, given https://github.com/whatwg/html/pull/3354#issuecomment-358604504, it seems maybe you already agree the cases other than `main` in `section` or `form` aren’t worth spending more time together on — so that is, for those other cases, the spec should define those as non-conforming (as @annevk wrote it already in this PR). Right?

Thanks for doing all that research! Are your tables for just the parent element so that `<body><form><div><main>` would could in the `<div>` category?

In any case, your numbers to confirm that it's really only `section` and `form` that are worth worrying about, as the rest makes less sense and is less common.

> So I dunno maybe that httparchive data needs to doublechecked to see why it unexpectedly shows `main` in `section` being used only 1.3 times as often as `main` in `form`.

Bugs are certainly possible, but shouldn't we expect rather different results from these two ways of measuring, just like how Chrome's use counters can show high usage for something that's still very hard to find in httparchive? For the checker the bias is people who tried to get it right. Since the `<form>` bits seem to be from an ASP framework, maybe that's just used a lot and people can't control the all of the markup and thus don't bother to validate?

So, should we allow `form` ancestors? I think it will look a bit curios that we allowed it but not for example `li`. @stevefaulkner, would it make sense to allow it only if the form has no accessible name, and what does that map to? Having a `name` attribute?

annevk commented:
@foolip per feedback from @AliceWonderMiscreations I think we should allow all main elements to be hidden, if any are present, which would include that case.

foolip commented:
I certainly support that change, but I was wondering about the case where an ancestors has the hidden attribute. The spec as written requires it to be on the main elements.

annevk commented:
Oh I see, I'd rather not go there and keep things as simple as we can.

stevefaulkner commented:
@foolip 
> would it make sense to allow it only if the form has no accessible name, and what does that map to? Having a name attribute?
Makes sense. In this case accessible name is provided from `aria-label/aria-labelledby` with fallback to `title` attribute if present and neither of previous are present.

sideshowbarker commented:
> I was wondering about the case where an ancestors has the hidden attribute.
>> Oh I see, I'd rather not go there and keep things as simple as we can.

Me too. At least unless/until we get feedback from users who are frustrated by it being that way (e.g., bug reports against the HTML checker).    

> @sideshowbarker, at what point does it just start to get annoying to explain in the error messages for the validator?

Well for the record here and since you asked, I don’t think it’d not be any more annoying to explain in a checker message than what  the current message already says, which is just:

> Error: A document must not include more than one visible `main` element.

Seems like I wouldn’t even need to change that message at all — since it already doesn’t what element the `hidden` attribute happens to be on.

And also as far as implementing “the case where an ancestors has the hidden attribute” on the checker backend, it wouldn’t be much additional work over what’s already implemented. (The existing check is already written in custom Java code rather than being expressed in the RelaxNG schema.)

annevk commented:
@foolip @stevefaulkner I thought there was agreement that allowing `<form>` was fine because of the ASP.NET pattern? If we require something of that `<form>` that would basically not help existing content. I also don't see why we'd equate `form` with `li`.

stevefaulkner commented:
@annevk @foolip 
>I thought there was agreement that allowing <form> was fine because of the ASP.NET pattern?

yes

>If we require something of that <form> that would basically not help existing content.

i thought @foolip meant disallowing only when the `form` element had an accessible name, which i would be surprised if it it effects any of the asp.net cases, but am fine with not restricting, as its likely that it will be caught by nesting restrictions anyway. Yeah I am unsure about the `li` connection as it is seriously unlikely that it would be used to wrap the main content of a document.

sideshowbarker commented:
> Are your tables for just the parent element so that `<body><form><div><main>` would could in the `<div>` category?

Not just for the parent element but instead for any ancestor. So it records a count for `<body><form><div><main>` in three places: as part of the total 47684, for `main` in `form` as well as for `main` in `div`.

> For the checker the bias is people who tried to get it right.

I kind of wish that were the case, but for better or worse I think the reality is that most documents which get fed to the checker are not necessarily from people trying to get it right. Whenever a grab a random URL from the checker logs, I find the document at the URL typically has many markup errors and many CSS — often literally dozens of both.

I think in part the circumstance behind that fact is, a lot of URLs are getting sent to the checker by some kind of automatic means not be the authors who wrote them but instead … in some other weird way for some odd reason(s).

And that’s the situation even after having set up a lot of filtering on the checker backend to block certain bad-behaving browser extensions and other things that are clearly junk requests.

So I think it’s not necessarily safe to assume much at all on scale about the average usage pattern for the checker. I think for every document it processes that’s submitted from an author/developer trying to do the right thing, there’s at least one document it processes that’s a unbelievably bad mess of really horrible markup and CSS.

annevk commented:
I made it to allow only "form without accessible name" as that seems reasonable. I also updated how I defined the hierarchy restriction to make it more similar to what we do for `area` as that avoids a lot of duplication throughout the document. I added "Alice Wonder" to the acknowledgments section (@AliceWonderMiscreations let me know if you prefer a different name there).

I think this is ready for a fresh round of reviews as I've addressed all the nits I found myself.

foolip commented:
OK, so let's leave the hidden attribute ancestor stuff alone, it'd be a pretty straightforward tweak later.

foolip commented on source:
```diff
@@ -19197,15 +19200,17 @@ included with Exhibit B.
 
-  <p class="note">While there is no restriction as to the number of <code>main</code> elements in a
-  document, web developers are encouraged to stick to a single element.</p>
+  <p>A <dfn>hierarchically correct <code>main</code> element</dfn> is one whose ancestor elements
```
This is great, simplified a lot of the other bits. Thanks!

annevk commented:
I'll land this next Tuesday, barring any objections. It'd be great to get explicit confirmation from @AliceWonderMiscreations on https://github.com/whatwg/html/pull/3354#issuecomment-358946485 before then though.

Also, thanks everyone for the help and my apologies about the delay in getting this done and all the angst it has caused over the intervening years. (To be clear, I realize more needs to be done and I will focus on that next.)

aardrian commented:
@iandevlin:
> @stevefaulkner Yes, ASP.NET requires everything to be wrapped in a `<form>` because the entire framework revolves around POST commands.

I thought that mostly disappeared in ASP.NET. If Microsoft is moving away / has moved away from it, that may not be sufficient justification.

AliceWonderMiscreations commented:
#3354 is fine way to reference me, thank you.

sideshowbarker commented:
The HTML checker now also implements the “hierarchically correct” restriction; https://github.com/validator/validator/commit/8024cd861c666890795b897d42c457666b598cb5

JaninaSajka commented:
Awesome! Thanks, Mike!

Janina

Michael[tm] Smith writes:
> The HTML checker now also implements the “hierarchically correct” restriction; https://github.com/validator/validator/commit/8024cd861c666890795b897d42c457666b598cb5
> 
> -- 
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/whatwg/html/pull/3354#issuecomment-362102434

-- 

Janina Sajka

Linux Foundation Fellow
Executive Chair, Accessibility Workgroup:	http://a11y.org

The World Wide Web Consortium (W3C), Web Accessibility Initiative (WAI)
Chair, Accessible Platform Architectures	http://www.w3.org/wai/apa
